moquette-0.16/broker/src/main/java/io/moquette/interception/BrokerInterceptor.java

```diff
@@ -0,0 +1,196 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * The Apache License v2.0 is available at
+ * http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.moquette.interception;
+
+import io.moquette.BrokerConstants;
+import io.moquette.interception.messages.*;
+import io.moquette.broker.config.IConfig;
+import io.moquette.broker.subscriptions.Subscription;
+import io.netty.handler.codec.mqtt.MqttConnectMessage;
+import io.netty.handler.codec.mqtt.MqttPublishMessage;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import static io.moquette.logging.LoggingUtils.getInterceptorIds;
+
+/**
+ * An interceptor that execute the interception tasks asynchronously.
+ */
+public final class BrokerInterceptor implements Interceptor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BrokerInterceptor.class);
+    private final Map<Class<?>, List<InterceptHandler>> handlers;
+    private final ExecutorService executor;
+
+    private BrokerInterceptor(int poolSize, List<InterceptHandler> handlers) {
+        LOG.info("Initializing broker interceptor. InterceptorIds={}", getInterceptorIds(handlers));
+        this.handlers = new HashMap<>();
+        for (Class<?> messageType : InterceptHandler.ALL_MESSAGE_TYPES) {
+            this.handlers.put(messageType, new CopyOnWriteArrayList<InterceptHandler>());
+        }
+        for (InterceptHandler handler : handlers) {
+            this.addInterceptHandler(handler);
+        }
+        executor = Executors.newFixedThreadPool(poolSize);
+    }
+
+    /**
+     * Configures a broker interceptor, with a thread pool of one thread.
+     *
+     * @param handlers InterceptHandlers listeners.
+     */
+    public BrokerInterceptor(List<InterceptHandler> handlers) {
+        this(1, handlers);
+    }
+
+    /**
+     * Configures a broker interceptor using the pool size specified in the IConfig argument.
+     * @param props configuration properties.
+     * @param handlers InterceptHandlers listeners.
+     *
+     */
+    public BrokerInterceptor(IConfig props, List<InterceptHandler> handlers) {
+        this(Integer.parseInt(props.getProperty(BrokerConstants.BROKER_INTERCEPTOR_THREAD_POOL_SIZE, "1")), handlers);
+    }
+
+    /**
+     * Shutdown graciously the executor service
+     */
+    public void stop() {
+        LOG.info("Shutting down interceptor thread pool...");
+        executor.shutdown();
+        try {
+            LOG.info("Waiting for thread pool tasks to terminate...");
+            executor.awaitTermination(10L, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)